### ISSUE TITLE
[Feature request] Add a way to act on warning "X has changed before removal"

### ISSUE THREAD

**Issue by scambier** [lucaong/minisearch#176]
Context: I'm using Minisearch to build a search plugin for the note-taking app Obisidian. To make it as fast as possible, I'm reloading Minisearch from cached data when the application boots up.

Since bugs happen, the cached index and cached files can become desynced, and the message `MiniSearch: document with ID xyz has changed before removal` appears. Unfortunately, we can't act on this, the errors pile up, and eventually the search index becomes unusable or corrupted.

I think a new callback field on the `Options` object would be a simple and efficient solution. It could be called when the `console.warn` is shown, so library users could manage it effectively.

**Comment by lucaong** [comment]
Hi @scambier ,
that makes sense. Let me think how to best do that. One way would be to have an option to throw an error in this case, instead of a warning.

**Comment by scambier** [comment]
Throwing an error makes sense but - for my use-case - I'd rather just warn the user, set a flag somewhere, and trigger a reindex for later (i.e. at next boot).
When the first warning pops up, it's usually relatively benign and affects a small subset of documents. I'd rather not kill the search engine outright _while it is being used_, and impose a (sometimes long) reindex when it is still 99% good.

**Comment by lucaong** [comment]
I am thinking to implement this in a way that is not overly specific to this use case. My idea is to add an optional configuration to specify a `logger` callback function:

```javascript
const miniSearch = new MiniSearch({
  fields: [/* ... */],
  logger: (level, message, code) => { /* logger implementation */ }
})
```

If `logger` is not specified, everything works like currently. Otherwise, the `logger` function is called any time `MiniSearch` needs to log some message or warning. At the moment the warning you mention is the only one, but this opens up the possibility to log more stuff in the future, maybe subject to a `logLevel` setting, and support custom loggers instead of relying on `console`.

The arguments of the callback would be the log level (`debug`, `info`, `warn`, `error`, etc.), the log message, and an optional `code` which can identify the specific use case, so one does not have to rely on matching the message content. In the case of the warning you mention, it could be `version_conflict`.

For reference, the default implementation of `logger` that maintains the current behavior would be:

```javascript
const defaultLogger = (level, message, _code) => {
  console[level](message)
}
```

Would this solve your case @scambier ?

**Comment by scambier** [comment]
Perfect 👌

**Comment by lucaong** [comment]
The pull request is merged, I will soon publish a new release

**Comment by lucaong** [comment]
Released on NPM as part of `v5.1.0`

**Comment by lucaong** [comment]
@scambier just a heads up: the newly released `v6.0.0-beta.1` (soon to be `v6.0.0`) introduces some new features that might make your case simpler and cleaner.

In particular, the new `discard` and `replace` methods allow removing a document by ID and replacing a document with a new version, without providing the original version. This wouldn’t suffer from the issue of the cached version being out of sync. You can read about them in [the documentation](https://lucaong.github.io/minisearch/classes/_minisearch_.minisearch.html).

**Comment by scambier** [comment]
oh my, that's awesome. Thank you for this!
